diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 7ab2319..333e325 100644
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -151,9 +151,11 @@ add_custom_command(OUTPUT ${SPIRV_TOOLS_BUILD_VERSION_INC}
    COMMENT "Update build-version.inc in the SPIRV-Tools build directory (if necessary).")
 # Convenience target for standalone generation of the build-version.inc file.
 # This is not required for any dependence chain.
-add_custom_target(spirv-tools-build-version
-   DEPENDS ${SPIRV_TOOLS_BUILD_VERSION_INC})
-set_property(TARGET spirv-tools-build-version PROPERTY FOLDER "SPIRV-Tools build")
+if (NOT IOS)
+  add_custom_target(spirv-tools-build-version
+    DEPENDS ${SPIRV_TOOLS_BUILD_VERSION_INC})
+  set_property(TARGET spirv-tools-build-version PROPERTY FOLDER "SPIRV-Tools build")
+endif()
 
 list(APPEND PCH_DEPENDS
 	${CORE_TABLES_HEADER_INC_FILE}
@@ -338,8 +340,11 @@ function(spirv_tools_default_target_options target)
   )
   set_property(TARGET ${target} PROPERTY FOLDER "SPIRV-Tools libraries")
   spvtools_check_symbol_exports(${target})
-  add_dependencies(${target}
-    spirv-tools-build-version core_tables extinst_tables)
+  if (IOS)
+    add_dependencies(${target} core_tables extinst_tables)
+  else ()
+    add_dependencies(${target} spirv-tools-build-version core_tables extinst_tables)
+  endif()
 endfunction()
 
 if (SPIRV_TOOLS_BUILD_SHARED)
