diff --git a/library/common.h b/library/common.h
index 50f2a29..372e80c 100644
--- a/library/common.h
+++ b/library/common.h
@@ -19,10 +19,10 @@
 #include <stdint.h>
 #include <stddef.h>
 
-#if defined(__ARM_NEON)
+#if defined(MBEDTLS_PLATFORM_IS_WINDOWS_ON_ARM64)
 #include <arm_neon.h>
 #define MBEDTLS_HAVE_NEON_INTRINSICS
-#elif defined(MBEDTLS_PLATFORM_IS_WINDOWS_ON_ARM64)
+#else
 #include <arm64_neon.h>
 #define MBEDTLS_HAVE_NEON_INTRINSICS
 #endif
