diff --git a/CMakeLists.txt b/CMakeLists.txt
index fcb5417..f8f71e3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -109,8 +109,7 @@ if(NOT QUAZIP_QT_ZLIB_USED)
 
         set(QUAZIP_LIB_LIBRARIES ${QUAZIP_LIB_LIBRARIES} ${ZLIB_LIBRARY})
     else()
-        find_package(ZLIB REQUIRED)
-        set(QUAZIP_LIB_LIBRARIES ${QUAZIP_LIB_LIBRARIES} ZLIB::ZLIB)
+        set(QUAZIP_LIB_LIBRARIES ${QUAZIP_LIB_LIBRARIES} z)
     endif()
 endif()
 
