From d46675fbb61eb6d51e478023ce4075e545ad4cfd Mon Sep 17 00:00:00 2001
From: crueter <crueter@eden-emu.dev>
Date: Sat, 30 May 2026 21:11:55 -0400
Subject: [PATCH] use ccache

---
 CMakeLists.txt               |  1 -
 cmake/ConfigureOpenSSL.cmake | 12 +++---------
 2 files changed, 3 insertions(+), 10 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index d6c998c..d943c59 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,7 +20,6 @@ include(ProcessorCount)
 include(cmake/CheckBuildMismatches.cmake)
 include(cmake/ConfigureOpenSSL.cmake)
 include(cmake/DetectTargetPlatform.cmake)
-include(cmake/FetchOpenSSL.cmake)
 include(cmake/FindVcvarsall.cmake)
 
 # Custom options
diff --git a/cmake/ConfigureOpenSSL.cmake b/cmake/ConfigureOpenSSL.cmake
index be8d651..c186c16 100644
--- a/cmake/ConfigureOpenSSL.cmake
+++ b/cmake/ConfigureOpenSSL.cmake
@@ -77,15 +77,9 @@ function(apply_ccache FILE)
         )
     endif()
 
-    if(OPENSSL_USE_CCACHE)
-        find_program(CCACHE ccache)
-
-        if(NOT CCACHE)
-            return()
-        endif()
-
+    if(USE_CCACHE)
         file(READ ${FILE} MAKEFILE)
-        string(REPLACE "\nCC=" "\nCC=ccache " MAKEFILE "${MAKEFILE}")
+        string(REPLACE "\nCC=" "\nCC=${CCACHE_BINARY} " MAKEFILE "${MAKEFILE}")
 
         if(MSVC)
             string(REPLACE "/Zi /Fdossl_static.pdb " "" MAKEFILE "${MAKEFILE}")
